src/main/java/network/brightspots/rcv/ResultsWriter.java

```diff
@@ -87,49 +89,131 @@ class ResultsWriter {
-  static String getOutputFilePath(
-      String outputDirectory,
-      String outputType,
-      String timestampString,
-      String sequentialTabulationId) {
-    outputType = sanitizeStringForOutput(outputType);
-    String fileName =
-        String.format(
-            "%s_%s%s",
-            timestampString, outputType, sequentialSuffixForOutputPath(sequentialTabulationId));
-    return Paths.get(outputDirectory, fileName).toAbsolutePath().toString();
+  public static class ResultFile {
+    // Since sanitizing Slice IDs can cause filename collisions, ensure each non-sanitized
+    // Slice ID is given a unique sanitized name.
+    private static final Dictionary<String, String> sliceIdToUniqueSanitizedId = new Hashtable<>();
+    // This is a set of the values in sanitizerCollisionResolution to ensure there are no collisions
+    // in the resulting sanitized names.
+    private static final HashSet<String> uniqueSanitizedIds = new HashSet<>();
+    private final ResultType resultType;
+    private final boolean isSlice;
+    private final TabulateBySlice slice;
+    private final String sliceId;
+
+    ResultFile(ResultType resultType) {
+      this.resultType = resultType;
+      this.isSlice = false;
+      slice = null;
+      sliceId = null;
+    }
+
+    ResultFile(ResultType resultType, TabulateBySlice slice, String sliceId) {
+      this.resultType = resultType;
+      this.isSlice = true;
+      this.slice = slice;
+      this.sliceId = sliceId;
+    }
+
+    // getPath helper without the modifier argument
+    public Path getPath(String directory, String prefix, Integer sequentialId) {
+      return getPath(directory, prefix, null, sequentialId);
+    }
+
+    // The filename is an underscore-separated string containing all components needed
+    // to create a unique output file. The modifier is likely only used in tests, where
+    // it is set to "expected" to distinguish expected output files from actual output files.
+    public Path getPath(String directory, String prefix, String modifier, Integer sequentialId) {
+      List<String> parts = new ArrayList<>();
+      parts.add(prefix);
+      if (sequentialId != null) {
+        parts.add(sequentialId.toString());
+      }
+      if (modifier != null) {
+        parts.add(modifier);
+      }
+      if (isSlice) {
+        assert slice != null;
+        parts.add(sanitizeSliceWithoutCollisions(sliceId));
+        parts.add(slice.toLowerString());
+      }
+      parts.add(resultType.getBasename());
+
+      String filenameWithoutExt = String.join("_", parts);
+      return Path.of(directory, "%s.%s".formatted(filenameWithoutExt, resultType.getExtension()));
+    }
+
+    private String sanitizeSliceWithoutCollisions(String sliceId) {
+      String previousSanitizedSliceId = sliceIdToUniqueSanitizedId.get(sliceId);
+      String sanitizedSliceId;
+
+      if (previousSanitizedSliceId != null) {
+        sanitizedSliceId = previousSanitizedSliceId;
```

Comment:
Unfortunately not -- something we failed to mention, but early returns are not allowed for compliance reasons.

Reply:
Wait, what? That's wild to me. If possible, can you provide the source for that statement? Just for my own education.

Reply:
Here's all I know: https://github.com/BrightSpots/rcv/pull/727#issuecomment-1610203649

@yezr do you know the vvsg source referred to there?

Reply:
Ah, found it. This PDF is linked in the contributor guide. See p94, very last bullet point: https://github.com/BrightSpots/rcv/blob/develop/reference/VVSG/VVSG-2005.1.0.VOL.1.pdf

Reply:
Wow, thank you! I will spend some time with this document get more familiar with these requirements. Early return whenever possible has been drilled so deep into my brain, I'll have a hard time letting go of it!

Reply:
It looks like the VVSG 2.0, released in 2021, is significantly more relaxed on this subject. I can't find any reference to single exit requirements. Are there any plans for this project to move to the 2.0 guidlines in the future? https://www.eac.gov/sites/default/files/TestingCertification/Voluntary_Voting_System_Guidelines_Version_2_0.pdf

Reply:
We're planning to get tested to VVSG 2.0 standards once RCTab v2.0 is completed. VVSG 2.0 rules were adopted in February 2021. 12 months after new rules are passed, all voting system have to use those rules. The EAC gave some extra time for compliance so as of late 2023, all new voting systems must use VVSG 2.0.  

The Election Assistance Commission ([EAC](eac.gov)) promulgates the VVSG rules and has [test assertions](https://www.eac.gov/sites/default/files/2023-07/VVSG_2.0_Test_Assertions_v1.3.pdf). Before testing I'll go line by line through them to ensure RCTab satisfies each one. Hoping to start that process soon since we're nearing completion of the 2.0 work. Section 2 of VVSG 2.0 contains a lot of prescriptions for code quality and programming guidelines. 

I don't know that there is an easy diff of the the latest VVSG to earlier versions. I could use a tool like that to see if there are any other coding conventions changed like this relaxed early return requirement. They have [this major changes document](https://www.eac.gov/sites/default/files/TestingCertification/VVSG_2_Major_Updates.pdf) but it's at a way higher level than individual coding conventions. Section 2.1-C in VVSG 2.0 requires that "published, credible" coding conventions be used but doesn't specify specific conventions. 

Like Alyssa I can't find anything explicit against early returns in VVSG 2.0. I will put an explicit check in my Todo list when going line by line through the VVSG 2.0 requirements to see if it has explicit guidelines but otherwise I think we are ok to simplify this with an early return if we want to. 

Also FYI a lot of state requirements say "use the federal requirements." So the EAC move to VVSG 2.0 also means a lot of states move to VVSG 2.0. You can see each state's certification strictness in [this EAC document](https://www.eac.gov/sites/default/files/2023-08/State%20Requirements%20for%20Certification%202023.pdf).

Reply:
I also cannot find this rule is VVSG 2.0, and if it's truly gone, I'd love to make a pass through the codebase to clean it up and early-return as often as is needed. I agree with Alyssa, it makes the code a lot simpler to read and understand if we can. I do think it should be a separate task to comb through the codebase rather than do it here, though.

Reply:
I made a note in my task for reviewing the VVSG 2.0 requirements to verify if this prohibition still exists. If I can confirm that it no longer exists I'll create a github issue to make the pass through the codebase that Armin suggested.